lib/wallet/src/Cardano/Wallet/DB/Sqlite/Migration/NewSubmissionStore.hs

```diff
@@ -0,0 +1,185 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Wallet.DB.Sqlite.Migration.NewSubmissionStore
+    ( createDatabase) where
+
+import Prelude
+
+import Cardano.Wallet.DB.Sqlite.Schema
+    ( LocalTxSubmission (..)
+    , TxCollateral (..)
+    , TxCollateralOut (..)
+    , TxCollateralOutToken (..)
+    , TxIn (..)
+    , TxMeta (..)
+    , TxOut (..)
+    , TxOutToken (TxOutToken)
+    , TxWithdrawal (TxWithdrawal)
+    , Wallet (..)
+    , migrateAll
+    )
+import Cardano.Wallet.DB.Sqlite.Types
+    ( BlockId (..), TxId (..) )
+import Cardano.Wallet.DB.Store.Meta.Model
+    ( TxMetaHistory (TxMetaHistory) )
+import Cardano.Wallet.DB.Store.Submissions.Model
+    ( TxLocalSubmissionHistory (TxLocalSubmissionHistory) )
+import Cardano.Wallet.DB.Store.Transactions.Model
+    ( TxRelation (TxRelation), TxSet (TxSet) )
+import Cardano.Wallet.DB.Store.Wallets.Model
+    ( MetasAndSubmissionsHistory, TxWalletsHistory )
+import Cardano.Wallet.DB.Store.Wallets.Store
+    ( mkStoreTxWalletsHistory )
+import Cardano.Wallet.Primitive.Types.RewardAccount
+    ( RewardAccount (..) )
+import Cardano.Wallet.Primitive.Types.TokenPolicy
+    ( TokenName (..), TokenPolicyId (..) )
+import Cardano.Wallet.Primitive.Types.TokenQuantity
+    ( TokenQuantity (..) )
+import Cardano.Wallet.Primitive.Types.Tx.SealedTx
+    ( mockSealedTx )
+import Control.Monad.Logger
+    ( runStderrLoggingT )
+import Control.Monad.Morph
+    ( MFunctor (hoist) )
+import Control.Monad.Trans
+    ( MonadIO (liftIO), MonadTrans (lift) )
+import Control.Monad.Writer
+    ( MonadWriter (tell), execWriter )
+import Crypto.Hash
+    ( hash )
+import Data.ByteString.Char8
+    ( ByteString )
+import Data.DBVar
+    ( Store (..) )
+import Data.Map.Strict
+    ( Map )
+import Data.Time
+    ( getCurrentTime )
+import Database.Persist.Sql
+    ( SqlPersistT, insert_, runMigration )
+import Database.Persist.Sqlite
+    ( runSqlite )
+
+import qualified Cardano.Wallet.Primitive.Types as W
+import qualified Cardano.Wallet.Primitive.Types.Address as W
+import qualified Cardano.Wallet.Primitive.Types.Coin as W
+import qualified Cardano.Wallet.Primitive.Types.Hash as W
+import qualified Cardano.Wallet.Primitive.Types.Tx.TxMeta as W
+    ( Direction (Outgoing), TxStatus (..) )
+
+createDatabase :: IO ()
+createDatabase = runStderrLoggingT $ runSqlite "before_submission.sqlite" $ do
+    runMigration migrateAll
+    liftIO mkWalletEntity >>= insert_
+    hoist (lift . lift . lift) storeAWallet
```

Comment:
```suggestion
    mapReaderT liftIO storeAWallet
```

I didn't type check this, but I think it should work. 🤔

* `mapReaderT` specializes `hoist`
* `liftIO` automatically combines the `lift`s

Reply:
accepted